vsp.go

```diff
@@ -57,65 +61,325 @@ func (mw *MultiWallet) NewVSPClient(vspHost string, walletID int, purchaseAccoun
+	networkBackend, err := vsp.w.Internal().NetworkBackend()
 	if err != nil {
 		return err
 	}
 
-	_, err = v.w.Internal().PurchaseTickets(v.ctx, networkBackend, request)
+	_, err = vsp.w.Internal().PurchaseTickets(vsp.ctx, networkBackend, request)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
+// StartTicketBuyer starts the automatic ticket buyer.
+func (vsp *VSP) StartTicketBuyer(balanceToMaintain int64, passphrase []byte) error {
+	if balanceToMaintain < 0 {
+		return errors.New("Negative balance to maintain given")
+	}
+
+	vsp.w.cancelAutoTicketBuyerMux.Lock()
+	cancelFunc := vsp.w.cancelAutoTicketBuyer
+	vsp.w.cancelAutoTicketBuyerMux.Unlock()
+	if cancelFunc != nil {
+		return errors.New("Ticket buyer already running")
+	}
+
+	ctx, cancel := vsp.w.shutdownContextWithCancel()
+
+	vsp.w.cancelAutoTicketBuyerMux.Lock()
+	vsp.w.cancelAutoTicketBuyer = cancel
+	vsp.w.cancelAutoTicketBuyerMux.Unlock()
+
+	if len(passphrase) > 0 && vsp.w.IsLocked() {
+		err := vsp.w.UnlockWallet(passphrase)
+		if err != nil {
+			return translateError(err)
+		}
+		defer vsp.w.LockWallet()
+	}
+
+	go func() {
+		log.Infof("Running ticket buyer on Wallet: %s", vsp.w.Name)
+
+		err := vsp.runTicketBuyer(ctx, balanceToMaintain, passphrase)
+		if err != nil {
+			if ctx.Err() != nil {
+				log.Errorf("[%d] TicketBuyer instance canceled, account number: %s", vsp.w.ID, vsp.w.Name)
+			}
+			log.Errorf("[%d] Ticket buyer instance errored: %v", vsp.w.ID, err)
+		}
+
+		vsp.w.cancelAutoTicketBuyerMux.Lock()
+		vsp.w.cancelAutoTicketBuyer = nil
+		vsp.w.cancelAutoTicketBuyerMux.Unlock()
+	}()
+
+	return nil
+}
+
+// runTicketBuyer executes the ticket buyer. If the private passphrase is incorrect,
+// or ever becomes incorrect due to a wallet passphrase change, runTicketBuyer exits with an
+// errors.Passphrase error.
+func (vsp *VSP) runTicketBuyer(ctx context.Context, balanceToMaintain int64, passphrase []byte) error {
+	if len(passphrase) > 0 && vsp.w.IsLocked() {
+		err := vsp.w.UnlockWallet(passphrase)
+		if err != nil {
+			return translateError(err)
+		}
+	}
+
+	c := vsp.w.Internal().NtfnServer.MainTipChangedNotifications()
+	defer c.Done()
+
+	ctx, outerCancel := context.WithCancel(ctx)
+	defer outerCancel()
+	var fatal error
+	var fatalMu sync.Mutex
+
+	var nextIntervalStart, expiry int32
+	var cancels []func()
+	for {
+		select {
+		case <-ctx.Done():
+			defer outerCancel()
+			fatalMu.Lock()
+			err := fatal
+			fatalMu.Unlock()
+			if err != nil {
+				return err
+			}
+			return ctx.Err()
+		case n := <-c.C:
+			if len(n.AttachedBlocks) == 0 {
+				continue
+			}
+
+			tip := n.AttachedBlocks[len(n.AttachedBlocks)-1]
+			w := vsp.w.Internal()
+
+			// Don't perform any actions while transactions are not synced through
+			// the tip block.
+			rp, err := w.RescanPoint(ctx)
+			if err != nil {
+				return err
+			}
+			if rp != nil {
+				log.Debugf("Skipping autobuyer actions: transactions are not synced on Wallet: %s", vsp.w.Name)
```

Comment:
For logging consistency, use wallet name or ID consistently. If using wallet name, no need to capitalize wallet:
```suggestion
				log.Debugf("Skipping autobuyer actions: transactions are not synced on wallet: %s", vsp.w.Name)
```